src/arkode/arkode_mristep.c

```diff
@@ -2995,6 +4727,143 @@ void mriStepInnerStepper_PrintMem(MRIStepInnerStepper stepper, FILE* outfile)
   return;
 }
 
+/*---------------------------------------------------------------
+  Utility routines for MRIStep to serve as an MRIStepInnerStepper
+  ---------------------------------------------------------------*/
+
+/*------------------------------------------------------------------------------
+  mriStep_ApplyForcing
+
+  Determines the linear combination coefficients and vectors to apply forcing
+  at a given value of the independent variable (t).  This occurs through
+  appending coefficients and N_Vector pointers to the underlying cvals and Xvecs
+  arrays in the step_mem structure.  The dereferenced input *nvec should indicate
+  the next available entry in the cvals/Xvecs arrays.  The input 's' is a
+  scaling factor that should be applied to each of these coefficients.
+  ----------------------------------------------------------------------------*/
+
+void mriStep_ApplyForcing(ARKodeMRIStepMem step_mem, sunrealtype t,
+                          sunrealtype s, int* nvec)
+{
+  sunrealtype tau, taui;
+  int i;
+
+  /* always append the constant forcing term */
+  step_mem->cvals[*nvec] = s;
+  step_mem->Xvecs[*nvec] = step_mem->forcing[0];
+  (*nvec) += 1;
+
+  /* compute normalized time tau and initialize tau^i */
+  tau  = (t - step_mem->tshift) / (step_mem->tscale);
+  taui = tau;
+  for (i = 1; i < step_mem->nforcing; i++)
+  {
+    step_mem->cvals[*nvec] = s * taui;
+    step_mem->Xvecs[*nvec] = step_mem->forcing[i];
+    taui *= tau;
+    (*nvec) += 1;
+  }
+}
+
+/*------------------------------------------------------------------------------
+  mriStep_SetInnerForcing
+
+  Sets an array of coefficient vectors for a time-dependent external polynomial
+  forcing term in the ODE RHS i.e., y' = f(t,y) + p(t). This function is
+  primarily intended for using MRIStep as an inner integrator within another
+  [outer] instance of MRIStep, where this instance is used to solve a
+  modified ODE at a fast time scale. The polynomial is of the form
+
+  p(t) = sum_{i = 0}^{nvecs - 1} forcing[i] * ((t - tshift) / (tscale))^i
+
+  where tshift and tscale are used to normalize the time t (e.g., with MRIGARK
+  methods).
+  ----------------------------------------------------------------------------*/
+
+int mriStep_SetInnerForcing(ARKodeMem ark_mem, sunrealtype tshift,
+                            sunrealtype tscale, N_Vector* forcing, int nvecs)
+{
+  ARKodeMRIStepMem step_mem;
+  int retval;
+
+  /* access ARKodeMRIStepMem structure */
+  retval = mriStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return (retval); }
+
+  if (nvecs > 0)
+  {
+    /* enable forcing, and signal that the corresponding pre-existing RHS
+       vector is no longer current, since it has a stale forcing function */
+    if (step_mem->explicit_rhs)
+    {
+      step_mem->expforcing     = SUNTRUE;
+      step_mem->impforcing     = SUNFALSE;
+      step_mem->fse_is_current = SUNFALSE;
```

Comment:
I think I just realized why `fn_is_current` was being used in the `FullRHS` function. Since `fse_is_current` and `fsi_is_current` are only reset to false when the forcing is updated they can't be used to determine if the RHS needs to be re-evaluated between steps. I'll go back now and reply to my comments related to `fn_is_current` with new suggestions.

Reply:
Ok, I've fixed the suggestions related to `fn_is_current`. It looks like `fse_is_current` and `fsi_is_current` checks are always paired with `fn_is_current` checks and the`fs*_is_current` flags are only set to false when the forcing is updated, so I think those flags could be removed (with the addition of setting `fn_is_current` to false when the forcing is changed as suggested above). That said, the `fs*_is_current` flags have given me an idea for how to simplify the `FullRhs` logic and `fn_is_current` flags checks but that can be done in a later PR.

Reply:
Yes, this was my purpose with having additional flags -- the main ARKODE infrastructure can signal changes through `fn_is_current`, but MRIStep knows about its separate RHS functions, and can act appropriately based on the added information.

Reply:
With the current setup (where the `fs*_is_current` flags are not updated after a successful step), the values of these flags aren't really used because `fn_is_current` is always part of the RHS evaluation logic so it ends up determining if an evaluation is needed so `fs*_is_current` could be removed. Although there is a way we could better leverage these flags but it will require some changes to the ARKODE infrastructure beyond the scope of this PR.

Reply:
I agree that whenever `fn_is_current == SUNFALSE` then all slow RHS functions will be evaluated.  However, the idea in adding the `fs*_is_current` flags is that they allow the situation where `fn_is_current == SUNTRUE` but one of `fs*_is_current == SUNFALSE`.  

This could occur on the "intermediate" integrator in a telescopic MRI integration.  Here, assume the intermediate-scale MRIStep integrator has already computed a valid step and its FullRHS was called with `ARK_FULLRHS_END` mode.  Following that call, assume the "outer" integrator calls `mriStep_SetInnerForcing` on the intermediate integrator.  Here, the intermediate-scale MRIStep integrator must indicate that the storage location for its forcing terms is no longer accurate.

In this situation, I do not see how `fs*_is_current` could be removed without having to instead set `fn_is_current == SUNFALSE` inside `mriStep_SetInnerForcing`, thus inducing an extraneous call to whichever of `fse` or `fsi` *does not store* the external forcing data.

Reply:
When the outer integrator calls `SetInnterForcing` on the intermediate integrator the intermediate integrator's stored value of `fn` is no longer valid since the problem has changed and the forcing used to compute `fn` is now incorrect going forward. So `fn_is_current` should be set to false otherwise the stale data could be reused (unless a reset happens, which also sets `fn_is_current` to false). 

You're right, we could avoid a RHS call to the whichever of `fse` or `fsi` does not include the forcing but I think this will require some modifications to ARKODE so we can set the `fs*_is_current` flags to false after every successful step so we don't have to rely on `fn_is_current`.